lib/elixir/test/elixir/kernel_test.exs

```diff
@@ -247,6 +247,30 @@ defmodule KernelTest do
     assert_raise BadBooleanError, fn -> 0 or 1 end
   end
 
+  defp check_struct(arg) when is_struct(arg, Date), do: :struct_date
+  defp check_struct(arg) when is_struct(arg), do: :struct
+  defp check_struct(_arg), do: :unknown
+
+  test "is_struct/1" do
+    assert is_struct(%{}) == false
+    assert is_struct(%{}) == false
+    assert is_struct(%Macro.Env{}) == true
+    assert is_struct([]) == false
+    assert is_struct(%{__struct__: "foo"}) == false
+    assert check_struct(%Macro.Env{}) == :struct
+    assert check_struct([]) == :unknown
+    assert check_struct(%{}) == :unknown
+  end
+
+  test "is_struct/2" do
+    assert is_struct(%{}, Macro.Env) == false
+    assert is_struct(%Macro.Env{}, Macro.Env) == true
+    assert is_struct(%Macro.Env{}, DateTime) == false
+    assert is_struct([], Macro.Env) == false
```

Comment:
We need to check what happens when you don't pass an atom as second argument. Ideally we would want to raise badarg, but that means we would need to make the guard fail instead of returning false and that is not easy to emulate.

Any ideas on how we could emulate it? If we can't, then we will need to postpone is_struct/2 for now. :/

Reply:
> We need to check what happens when you don't pass an atom as second argument. Ideally we would want to raise badarg, but that means we would need to make the guard fail instead of returning false and that is not easy to emulate.
> 
> Any ideas on how we could emulate it? If we can't, then we will need to postpone is_struct/2 for now. :/

Why can't we just check for `is_atom(struct)` in https://github.com/elixir-lang/elixir/pull/9470/files#diff-e06d5ebc2907d951188e88476f85af51R2235 and if not raise `BadArgument`?
Should we care in the `:guard` for this in the guard context, since any function that raises is considered `false`?

Reply:
Because nothing but an atom is a valid struct name. So giving anything other, should be a bad argument. For example, is_function/2 raises when the second argument is not an integer.

Reply:
> Because nothing but an atom is a valid struct name. So giving anything other, should be a bad argument. For example, is_function/2 raises when the second argument is not an integer.

I know that, and in fact I run `is_function/2` before writing my comment, but what prevents us from running `is_atom(struct)` and raising a `BarArgument` error if struct is not an atom. Or am I missing something?

Reply:
@josevalim the only way I know to emulate such behaviour would be to:

```
(is_atom(struct) or hd([]))
and :erlang.map_get(:__struct__, map) == struct
```

Reply:
@hauleth yeah, good call. I thought about `length(:foo)` but I am really uncomfortable with said approach. It can throw off code analysis tools and what not. Let's skip `is_struct/2` for now.